contrib/pg_tde/src/pg_tde_event_capture.c

```diff
@@ -141,6 +143,96 @@ push_event_stack(const TdeDdlEvent *event)
 	MemoryContextSwitchTo(oldCtx);
 }
 
+static List *
+find_typed_table_dependencies(Oid typeOid)
+{
+	Relation	classRel;
+	ScanKeyData key[1];
+	TableScanDesc scan;
+	HeapTuple	tuple;
+	List	   *result = NIL;
+
+	classRel = table_open(RelationRelationId, AccessShareLock);
+
+	ScanKeyInit(&key[0],
+				Anum_pg_class_reloftype,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(typeOid));
+
+	scan = table_beginscan_catalog(classRel, 1, key);
+
+	while ((tuple = heap_getnext(scan, ForwardScanDirection)) != NULL)
+	{
+		Form_pg_class classform = (Form_pg_class) GETSTRUCT(tuple);
+
+		LockRelationOid(classform->oid, AccessShareLock);
+		result = lappend_oid(result, classform->oid);
+	}
+
+	table_endscan(scan);
+	table_close(classRel, AccessShareLock);
+
+	return result;
+}
+
+typedef enum
+{
+	ENC_MIX_UNKNOWN,
+	ENC_MIX_PLAIN,
+	ENC_MIX_ENCRYPTED,
+	ENC_MIX_MIXED,
+} EncryptionMix;
+
+/*
+ * Since ALTER TABLE can modify multiple due to inheritance or typed tables
+ * which can for example result in TOAST tables being created for some or all
+ * of the modified tables while the event trigger is only fired once we cannot
+ * rely on the event stack to make sure we get the correct encryption status.
+ *
+ * Our solution is to be cautious and only modify tables when all tables with
+ * storage are either encrypt or not encrypted. If there is a mix we will
+ * throw and error. The result of this is also used to properly inform the
+ * SMGR of the current encryption status.
+ */
+static EncryptionMix
+alter_table_encryption_mix(Oid relid)
+{
+	EncryptionMix enc = ENC_MIX_UNKNOWN;
+	Relation	rel;
+	List	   *children;
+	ListCell   *lc;
+
+	rel = relation_open(relid, NoLock);
+
+	if (RELKIND_HAS_STORAGE(rel->rd_rel->relkind))
+		enc = rel->rd_rel->relam == get_tde_table_am_oid() ? ENC_MIX_ENCRYPTED : ENC_MIX_PLAIN;
+
+	if (rel->rd_rel->relkind == RELKIND_COMPOSITE_TYPE)
```

Comment:
Is this really an one or the other? It seems like I can create an inherited table from a typed table.

Reply:
Yes, you can. And I think my code should account for that since the function is recursive and those two calls only fetch direct children.

Reply:
I see, for types we first check for the children of the type (tables), and then the children of the tables (possible inheritance). Yes, that should work, but maybe it would be better to have an explicit testcase about it.